### PR TITLE
Initial support for Qwen3.5 family

### PR DESCRIPTION
**Pull request by haok1402**
Adds training support for the Qwen3.5-MoE architecture, starting with Qwen3.5-35B-A3B. This is an initial cut: it covers the text tower only, and context parallelism is not yet implemented.

## What this adds

- **Model** (`pithtrain/models/qwen35_moe.py`): the Qwen3.5-MoE text model, implemented against the 5-stage DualPipeV layer protocol. It is a hybrid MoE: most layers use Gated DeltaNet (linear attention) and the rest use full softmax attention (output-gated GQA with partial rotary). Every layer is MoE, with routed experts plus a sigmoid-gated shared expert.
- **Operator** (`pithtrain/operators/gated_delta_rule.py`): wraps FLA's fused chunked gated delta rule as a `torch.library` custom op, so the linear-attention layers stay fullgraph torch.compile friendly. Adds `flash-linear-attention` as a dependency.
- **Checkpoint conversion** (`pithtrain/tasks/convert_checkpoint/qwen35_moe.py`): converts the released checkpoint to and from the runtime DCP format. The released model is vision-language, so only the text tower (`model.language_model.*` and `lm_head`) is converted; the vision tower and MTP weights are dropped. GPT-OSS detection is tightened to its per-expert bias key so the two fused-expert formats are not confused.
- **Examples and benchmarks**: pretrain, tokenize, and convert configs under `examples/.../qwen3.5-35b-a3b`, plus a pretraining setup with 2-node and 4-node configs under `benchmarks/pretraining/qwen3.5-35b-a3b`.
- **Tests** (`tests/test_fsdp.py`): extends the FSDP correctness ladder to the new model.

## Validation

- Single-GPU reference forward matches the 5-stage DualPipeV path.
- FSDP correctness ladder passes from pp1/ep1 up to pp2/ep2, with matching gradients.
- Checkpoint conversion round-trips byte for byte.
- End-to-end pretraining converges at pp2/ep8 across two nodes.

## Loaded-checkpoint correctness

Loading the converted release checkpoint and running a short job (pp2/ep8, sequence length 1536) starts at a cross-entropy loss around 2.4 on the very first step and holds there, rather than the high loss a randomly initialized model would show. This confirms the weight conversion and the forward path are correct end to end.

```
[rank0] layer_partition: 40 layers / 4 stages -> [10, 10, 10, 10]
[rank0] layer_partition: 40 layers / 4 stages -> [10, 10, 10, 10]
2026-06-22 18:28:45 | INFO | launch(cfg=PretrainLMCfg(distributed=DistributedCfg(pipeline_parallel_size=2, context_parallel_size=1, expert_parallel_size=8, timeout=datetime.timedelta(seconds=900), sharding_strategy='fsdp'), training=TrainingCfg(dataset=PosixPath('workspace/datasets/dclm-baseline/toktxt/qwen3.5'), sequence_length=1536, seed=1234, lr=1e-06, max_steps=25, micro_batch_size=1, global_batch_size=1024, optimizer=<function make_muon_optimizer at 0x134b422c8720>, scheduler=<function make_constant_scheduler at 0x134b422c8b40>, model=PosixPath('benchmarks/pretraining/qwen3.5-35b-a3b/model.json'), save_interval=None, save_location=PosixPath('workspace/checkpoints/qwen3.5-35b-a3b'), moe_load_balance_coef=0.001, moe_load_balance_type='global-batch', fp8_training='disabled', init_std=0.02, nsys_start=None, nsys_stop=None, memory_profile_start=None, memory_profile_stop=None, memory_profile_output=PosixPath('/home/haok/pith-train')), logging=LoggingCfg(wandb=None)))
2026-06-22 18:28:45 | INFO | Load checkpoint: workspace/checkpoints/qwen3.5-35b-a3b/torch-dcp/step-00000000
2026-06-22 18:28:52 | INFO | Load checkpoint: Elapsed min=7.4s, max=7.4s
2026-06-22 18:30:04 | INFO | step 00000001/00000025 | step-time 71.522 sec | cross-entropy-loss 2.4616 | load-balance-loss 1.066807 | learning-rate 1.000000e-06 | gradient-norm 355.1682 | tokens-per-second 21,991 | peak-gpu-memory 54.40 GB
2026-06-22 18:30:27 | INFO | step 00000002/00000025 | step-time 22.658 sec | cross-entropy-loss 2.4768 | load-balance-loss 1.064240 | learning-rate 1.000000e-06 | gradient-norm 373.1439 | tokens-per-second 69,417 | peak-gpu-memory 54.37 GB
2026-06-22 18:30:50 | INFO | step 00000003/00000025 | step-time 22.530 sec | cross-entropy-loss 2.4182 | load-balance-loss 1.072402 | learning-rate 1.000000e-06 | gradient-norm 360.1254 | tokens-per-second 69,812 | peak-gpu-memory 54.39 GB
2026-06-22 18:31:13 | INFO | step 00000004/00000025 | step-time 22.651 sec | cross-entropy-loss 2.3657 | load-balance-loss 1.066703 | learning-rate 1.000000e-06 | gradient-norm 337.9038 | tokens-per-second 69,440 | peak-gpu-memory 54.37 GB
2026-06-22 18:31:36 | INFO | step 00000005/00000025 | step-time 22.569 sec | cross-entropy-loss 2.4554 | load-balance-loss 1.070877 | learning-rate 1.000000e-06 | gradient-norm 268.8993 | tokens-per-second 69,692 | peak-gpu-memory 54.43 GB
```

### CODE REVIEW

**Comment by haok1402** [comment]
Mark as closed. I'll rebase and re-open to trigger the Claude Review workflow from #63
